Added FastChargeTimer Setting (get and set)

alexsee75 opened the pull request:
I have a project where a 10,000mAh battery is charging using Arduino 1010 and related platforms. Thermal constraints restrict the effective charging current to about 500mA which needs more than 8h (the default) to charge, yielding a charge fault (blinking CHRG LED). The simplest way to fix this was to add a setting for the charge timer which can be set up to 20h. The attached commit implements this - the set function rounds up since only four values are supported.

(I'll still try adding a heatsink to increase charging current, however even at 1A the charging time of 8h is too small)

I hereby contribute this code under the same license as the original code, or public domain if more convenient.

Best,
  Alex

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/arduino-libraries/Arduino_BQ24195?pullRequest=31) <br/>All committers have signed the CLA.

aentinger commented on src/BQ24195.cpp:
```diff
@@ -1195,6 +1202,68 @@ bool PMICClass::isBatteryInOverVoltage() {
+ * Input          : safety timer duration in hours (rounds up)
+ * Return         : 0 on Error, 1 on Success
+ *******************************************************************************/
+bool PMICClass::setFastChargeTimerSetting(float hours) {
```
```suggestion
bool PMICClass::setFastChargeTimerSetting(float const hours) {
```
Try and use const-correctness as much as possible :wink:

aentinger commented on src/BQ24195.cpp:
```diff
@@ -1195,6 +1202,68 @@ bool PMICClass::isBatteryInOverVoltage() {
+ *******************************************************************************/
+bool PMICClass::setFastChargeTimerSetting(float hours) {
+
+    int DATA = readRegister(CHARGE_TIMER_CONTROL_REGISTER);
```
```suggestion
    int const DATA = readRegister(CHARGE_TIMER_CONTROL_REGISTER);
```
Same here. Also I personally would be writing `DATA` lowercase, as all upper-case are reserve for constants usually.

aentinger commented on src/BQ24195.cpp:
```diff
@@ -1195,6 +1202,68 @@ bool PMICClass::isBatteryInOverVoltage() {
+        return 0;
+    }
+
+    byte mask = DATA & 0b11111001;
```
Here now I'd love to have a more expressive bitmask. You seem to be masing bytes 7, 1 and 2. Which meaning does those bytes have? I suggest using names for those bits and then rewrite using bitmasks, i.e.
```C++
enum CHARGE_TIMER_CONTROL_REGISTER
{
  BIT_NAME_7 = 7,
  BIT_NAME_2 = 2,
  BIT_NAME_1 = 1,
};

DATA &= ~((1<<BIT_NAME_7) | (1<<BIT_NAME_2) | (1<<BIT_NAME_1));
```

aentinger commented on src/BQ24195.cpp:
```diff
@@ -1195,6 +1202,68 @@ bool PMICClass::isBatteryInOverVoltage() {
+        return NAN;
+    }
+
+    byte mask = (DATA & 0b00000110) >> 1;
```
No need to shift right by 1, switch values must not start with zero.

aentinger reviewed (commented):
Thank you for your PR, looks solid :+1: . I've marked a couple of things that fall under "handicraft".

aentinger reviewed (changes requested):
See comments.

alexsee75 commented:
I've added the changes except for DATA which is uppercase in every function of this library - should be changed globally.
I've also added a writeRegister to disable the timer before setting the charging timer value, and reenabling it afterwards which is recommended by the BQ24195 docs.

I've thought long about whether to restore the previous EN_TIMER settings (disabled or enabled) or just set it afterwards - went with the latter, but there are probably arguments for doing it differently. E.g. I could check whether the charging timer is enabled - if not, can directly set the new charging timer value (saves one writeRegister) - if yes, disable it, set timer and reenable (as it is now). Not sure if it's an issue..

Best,
  Alex

aentinger reviewed (approved):
LGTM :+1: Thank you @alexsee75 :rocket:
